tools/schemapi/vega_expr.py

```diff
@@ -0,0 +1,1000 @@
+from __future__ import annotations
+
+import dataclasses
+import enum
+import keyword
+import re
+from collections import deque
+from inspect import getmembers
+from itertools import chain
+from pathlib import Path
+from textwrap import TextWrapper as _TextWrapper
+from textwrap import indent
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    ClassVar,
+    Iterable,
+    Iterator,
+    Literal,
+    Mapping,
+    Sequence,
+    overload,
+)
+from urllib import request
+
+from tools.markup import RSTParse, Token, read_ast_tokens
+from tools.markup import RSTRenderer as _RSTRenderer
+from tools.schemapi.schemapi import SchemaBase as _SchemaBase
+from tools.schemapi.utils import (
+    ruff_write_lint_format_str as _ruff_write_lint_format_str,
+)
+
+if TYPE_CHECKING:
+    import sys
+    from re import Match, Pattern
+
+    from mistune import BlockState
+
+    if sys.version_info >= (3, 11):
+        from typing import LiteralString, Self
+    else:
+        from typing_extensions import LiteralString, Self
+    from _typeshed import SupportsKeysAndGetItem
+
+__all__ = ["parse_expressions", "write_expr_module"]
+
+
+# NOTE: Urls/fragments
+VEGA_DOCS_URL: LiteralString = "https://vega.github.io/vega/docs/"
+EXPRESSIONS_DOCS_URL: LiteralString = f"{VEGA_DOCS_URL}expressions/"
+
+
+class Source(str, enum.Enum):
+    """Enumerations for ``expressions.md`` source files."""
+
+    LIVE = "https://raw.githubusercontent.com/vega/vega/main/docs/docs/expressions.md"
+    STATIC = "https://raw.githubusercontent.com/vega/vega/fb2e60274071033b4c427410ef43375b6f314cf2/docs/docs/expressions.md"
+    OLD = "https://raw.githubusercontent.com/vega/vega/ff98519cce32b776a98d01dd982467d76fc9ee34/docs/docs/expressions.md"
+
```

Comment:
## TODO
- Use release versions instead of commit hashes
- Investigate linking the `vega` version to https://github.com/vega/altair/blob/cabf1e6428bab107a26adc6c51993d3b2c5bf6f0/tools/generate_schema_wrapper.py#L49

Reply:
Provided a manual solution in [`d75792b` (#3600)](https://github.com/vega/altair/pull/3600/commits/d75792b15a029f1a1628728ef3403aadc30ab065)

So far I've only come up with [this](https://raw.githubusercontent.com/vega/vega-lite/refs/tags/v5.20.1/package.json) as a source for automating:

https://github.com/vega/vega-lite/blob/c7697dc293b606ba60c7e374dbf9a77226f91225/package.json#L137-L138

@mattijn I guess I have two questions:
1. Would manually syncing the `(vega|vega-lite)` versions seem reasonable to you?
2. Do you know of any simpler sources for getting the corresponding `vega` minimum version for a given `vega-lite` release?

I'm unsure whether using `package.json` is a good/bad idea, it was just the first source I found

Reply:
Tentative solution in:
- [build(DRAFT): Add vegalite_to_vega_version](https://github.com/vega/altair/pull/3600/commits/061b066e6a652d53e8cd94301439294c56338463)

Reply:
Manually syncing is fine by me. Adding a note in this [section](https://github.com/vega/altair/blob/main/NOTES_FOR_MAINTAINERS.md#updating-the-vega-lite-version) that we should update the `VEGA_VERSION` by eg inspecting the file you found or just the latest released Vega version.

Reply:
@mattijn I've settled on the solution here for now:
- [feat: Adds `vl_convert_to_vega_version`, remove `vegalite_to_vega_ver…`](https://github.com/vega/altair/pull/3600/commits/d95a5735c0ae784361529d7d50d558eee5dca7c3)

I'm going to open an issue on https://github.com/vega/vl-convert to see if we can later replace this with a public `vl_convert` function/attribute.
The result would be the same and depend on the same `const`, but would be less reliant on the structure of [vl-convert-vendor/src/main.rs](https://github.com/vega/vl-convert/blob/main/vl-convert-vendor/src/main.rs)

Reply:
Marking thread as resolved, follow this issue for updates

- https://github.com/vega/vl-convert/issues/191